### DISCUSSION TITLE
[Notifier] How to maintain previous state while new state is being built

### DISCUSSION THREAD

**Discussion by uny** [rrousselGit/riverpod#2151] [Q&A] [answered]
How to combine provider states with Notifier?
I would like to maintain its previous state.

Pattern 1.
```
class MyHomePageStateNotifier extends Notifier<MyHomePageState> {
  @override
  MyHomePageState build() {
    return MyHomePageState(
      counter: 0,
      title: ref.watch(titleProvider),
    );
  }
}
```
This will reset counter every time the titleProvider changes.

Pattern 2.
```
class MyHomePageStateNotifier extends Notifier<MyHomePageState> {
    try {
      return MyHomePageState(
        counter: state.counter,
        title: ref.watch(titleProvider),
      );
    } catch (exception) {
      return MyHomePageState(
        counter: 0,
        title: ref.watch(titleProvider),
      );
    }
  }
}
```
This will also reset counter because it causes `Bad state: Tried to read the state of an uninitialized provider` when titleProvider changes.

Pattern 3.
```
class MyHomePageStateNotifier extends Notifier<MyHomePageState> {
  @override
  MyHomePageState build() {
    ref.listen(titleProvider, (previous, next) {
      state = MyHomePageState(
        counter: state.counter,
        title: next,
      );
    });
    return const MyHomePageState(
      counter: 0,
      title: '',
    );
  }
}
```
Is `listen` the best practice? If so, should I close the ProviderSubscription returned by `listen` and how?

**Comment by rrousselGit** [comment, marked as answer]
You probably shouldn't have "title" in your state. Have the UI read titleProvider directly, and have your notifier focus on the counter.

By keeping those values split, there's no issue

**Reply by uny** [reply]
Thank you for your reply.
Putting aside the policy since this is a simplified example, does `Notifier` not support this use case?

**Reply by rrousselGit** [reply]
It's not about what's supposed or not. In the end you did it with Notifier using ref.listen
It's about how you design your code to use the most out of things like `ref.watch`


The solution I gave should apply to more advanced cases too. Splitting your provider into two or more is fine

**Reply by uny** [reply]
OK, I will try that.
